File: exercises/RB101_small_problems/medium_2/matching_parentheses.rb
=begin
# --------------------------- Problem
  Restate the problem:

  Write a method that takes a string as an argument, and returns true if
  all parentheses in the string are properly balanced, false otherwise.
  To be properly balanced, parentheses must occur in matching '(' and ')' pairs.

  Input: string
  Output: boolean (true/false)

# --------------------------- Test Cases + Rules

  Explicit:
    - count all '(' and ')' pairs
    - makesure it starts with '(' and ends with ')'
    - true if string contains no parentheses

  Implicit:
    - none identitified

  Test cases shown at the bottom

# --------------------------- Data Structure

  str -> array -> boolean

  Array: push parentheses to array and check conditions

# --------------------------- Scrapbook

# --------------------------- Algorithm

  /* given a string */
    - iterate through each char using array
      - if char is a parentheses, append to array
    - return true if:
      - number of '(' equals number of ')'
      - array starts with '(' and ends with ')'
      - array is empty
    - otherwise return false
    
=end

def balanced?(str)
  str.chars.each_with_object(arr = []) do |char, arr|
    arr << char if char == '(' || char == ')'
  end
  
  if arr.count('(') == arr.count(')') &&
    arr[0] == '(' && arr[-1] == ')' ||
    arr.empty?
    true
  else
    false
  end
end

p balanced?('What (is) this?') == true
p balanced?('What is) this?') == false
p balanced?('What (is this?') == false
p balanced?('((What) (is this))?') == true
p balanced?('((What)) (is this))?') == false
p balanced?('Hey!') == true
p balanced?(')Hey!(') == false
p balanced?('What ((is))) up(') == false

File: LS_215/02_String_and_Text_Processing/matchingParentheses.js
// Using for loop
function isBalanced(str) {
  let count = 0;

  for (let i = 0; i < str.length; i++) {
    if (str[i] === '(') count += 1;
    if (str[i] === ')') count -= 1;
    if (count < 0) return false;
  }

  return count === 0;
}

// Using forEach()
function isBalanced(str) {
  let count = 0;
  let unbalanced = false;

  [...str].forEach(char => {
    if (char === '(') count += 1;
    if (char === ')') count -= 1;
    if (count < 0) unbalanced = !unbalanced;
  });

  return unbalanced ? false : count === 0;
}


console.log(isBalanced('What (is) this?'));        // true
console.log(isBalanced('What is) this?'));         // false
console.log(isBalanced('What (is this?'));         // false
console.log(isBalanced('((What) (is this))?'));    // true
console.log(isBalanced('((What)) (is this))?'));   // false
console.log(isBalanced('Hey!'));                   // true
console.log(isBalanced(')Hey!('));                 // false
console.log(isBalanced('What ((is))) up('));       // false